### ISSUE TITLE
TPU performance increases (60x) when adding a dummy operation

### ISSUE THREAD

**Issue by bliep** [tensorflow/tensorflow#51485]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab
- TensorFlow version (use command below): 2.5.0
- device: TPU

**Describe the current behavior**
A model with 4 (or more) GRU layers and a Dense layer trains 60x faster if a dummy operation is inserted.

With the dummy operation the training time is around 2 **minutes** and 30 seconds per epoch:
<img width="699" alt="Screenshot 2021-08-13 at 11 27 29" src="https://user-images.githubusercontent.com/1833211/129337565-70ff96aa-f4d0-4edd-933b-f3f4b7af3ae8.png">

Without the dummy operation the training time is around 2 **hours**  and 30 mins per epoch:
<img width="628" alt="Screenshot 2021-08-13 at 11 26 40" src="https://user-images.githubusercontent.com/1833211/129337771-c570e0a7-5ad3-45fc-ae59-d0cdfcda9872.png">

**Describe the expected behavior**
The dummy operation should not have a large influence (and not be needed in the first place)

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1lbDTJp1JMWg5niM2Jtq2n2flWnkfLwpu?usp=sharing
Note the # dummy OP line, if this line is commented out the training loop runs ~60x slower.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by sushreebarsa** [comment]
@bliep Could you please take a look on the [link](https://www.tensorflow.org/guide/tpu) , [link1](https://cloud.google.com/tpu/docs/performance-guide) and let us know if it helps ? Thanks!

**Comment by bliep** [comment]
> @bliep Could you please take a look on the [link](https://www.tensorflow.org/guide/tpu) , [link1](https://cloud.google.com/tpu/docs/performance-guide) and let us know if it helps ? Thanks!

I have (previously) read the TPU guide and the performance guide, the issue is more involved. (so, no, the links do not help, sorry)

**Comment by sushreebarsa** [comment]
@bliep Thank you for the response! Could you please try on Colab using the latest version of TensorFlow 2.6.0 by  the command       ` !pip install tensorflow==2.6.0` and let us know if the issue still persists ? Thanks!

**Comment by bliep** [comment]
on 2.6.0 an error is thrown:

```
2.6.0
INFO:tensorflow:Clearing out eager caches
INFO:tensorflow:Clearing out eager caches
INFO:tensorflow:Initializing the TPU system: grpc://10.60.43.130:8470
INFO:tensorflow:Initializing the TPU system: grpc://10.60.43.130:8470
INFO:tensorflow:Finished initializing TPU system.
INFO:tensorflow:Finished initializing TPU system.
INFO:tensorflow:Found TPU system:
INFO:tensorflow:Found TPU system:
INFO:tensorflow:*** Num TPU Cores: 8
INFO:tensorflow:*** Num TPU Cores: 8
INFO:tensorflow:*** Num TPU Workers: 1
INFO:tensorflow:*** Num TPU Workers: 1
INFO:tensorflow:*** Num TPU Cores Per Worker: 8
INFO:tensorflow:*** Num TPU Cores Per Worker: 8
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:localhost/replica:0/task:0/device:CPU:0, CPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:localhost/replica:0/task:0/device:CPU:0, CPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:CPU:0, CPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:CPU:0, CPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:0, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:0, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:1, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:1, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:2, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:2, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:3, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:3, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:4, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:4, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:5, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:5, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:6, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:6, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:7, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:7, TPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU_SYSTEM:0, TPU_SYSTEM, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU_SYSTEM:0, TPU_SYSTEM, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:XLA_CPU:0, XLA_CPU, 0, 0)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:XLA_CPU:0, XLA_CPU, 0, 0)
All devices:  [LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:0', device_type='TPU'), LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:1', device_type='TPU'), LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:2', device_type='TPU'), LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:3', device_type='TPU'), LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:4', device_type='TPU'), LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:5', device_type='TPU'), LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:6', device_type='TPU'), LogicalDevice(name='/job:worker/replica:0/task:0/device:TPU:7', device_type='TPU')]
---------------------------------------------------------------------------
NotFoundError                             Traceback (most recent call last)
<ipython-input-1-76be564800d5> in <module>()
     39 with strategy.scope():
     40 
---> 41   model = create_model()
     42   model.summary()
     43   model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001), loss='mse')

30 frames
<ipython-input-1-76be564800d5> in create_model()
     23 def create_model():
     24   x = tf.keras.Input(shape=(n_seq_len, n_bins))
---> 25   g = tf.keras.layers.GRU(n_nodes, return_sequences=True)(x)
     26   g = tf.keras.layers.GRU(n_nodes, return_sequences=True)(g)
     27   g = tf.keras.layers.GRU(n_nodes, return_sequences=True)(g)

/usr/local/lib/python3.7/dist-packages/keras/layers/recurrent.py in __call__(self, inputs, initial_state, constants, **kwargs)
    657 
    658     if initial_state is None and constants is None:
--> 659       return super(RNN, self).__call__(inputs, **kwargs)
    660 
    661     # If any of `initial_state` or `constants` are specified and are Keras

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py in __call__(self, *args, **kwargs)
    975     if _in_functional_construction_mode(self, inputs, args, kwargs, input_list):
    976       return self._functional_construction_call(inputs, args, kwargs,
--> 977                                                 input_list)
    978 
    979     # Maintains info about the `Layer.call` stack.

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py in _functional_construction_call(self, inputs, args, kwargs, input_list)
   1113       # Check input assumptions set after layer building, e.g. input shape.
   1114       outputs = self._keras_tensor_symbolic_call(
-> 1115           inputs, input_masks, args, kwargs)
   1116 
   1117       if outputs is None:

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py in _keras_tensor_symbolic_call(self, inputs, input_masks, args, kwargs)
    846       return tf.nest.map_structure(keras_tensor.KerasTensor, output_signature)
    847     else:
--> 848       return self._infer_output_signature(inputs, args, kwargs, input_masks)
    849 
    850   def _infer_output_signature(self, inputs, args, kwargs, input_masks):

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py in _infer_output_signature(self, inputs, args, kwargs, input_masks)
    884           # overridden).
    885           # TODO(kaftan): do we maybe_build here, or have we already done it?
--> 886           self._maybe_build(inputs)
    887           inputs = self._maybe_cast_inputs(inputs)
    888           outputs = call_fn(inputs, *args, **kwargs)

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py in _maybe_build(self, inputs)
   2657         # operations.
   2658         with tf_utils.maybe_init_scope(self):
-> 2659           self.build(input_shapes)  # pylint:disable=not-callable
   2660       # We must set also ensure that the layer is marked as built, and the build
   2661       # shape is stored since user defined build functions may not be calling

/usr/local/lib/python3.7/dist-packages/keras/layers/recurrent.py in build(self, input_shape)
    575     if isinstance(self.cell, Layer) and not self.cell.built:
    576       with backend.name_scope(self.cell.name):
--> 577         self.cell.build(step_input_shape)
    578         self.cell.built = True
    579 

/usr/local/lib/python3.7/dist-packages/keras/utils/tf_utils.py in wrapper(instance, input_shape)
    257     if input_shape is not None:
    258       input_shape = convert_shapes(input_shape, to_tuples=True)
--> 259     output_shape = fn(instance, input_shape)
    260     # Return shapes from `fn` as TensorShapes.
    261     if output_shape is not None:

/usr/local/lib/python3.7/dist-packages/keras/layers/recurrent.py in build(self, input_shape)
   1796         regularizer=self.kernel_regularizer,
   1797         constraint=self.kernel_constraint,
-> 1798         caching_device=default_caching_device)
   1799     self.recurrent_kernel = self.add_weight(
   1800         shape=(self.units, self.units * 3),

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py in add_weight(self, name, shape, dtype, initializer, regularizer, trainable, constraint, use_resource, synchronization, aggregation, **kwargs)
    661         synchronization=synchronization,
    662         aggregation=aggregation,
--> 663         caching_device=caching_device)
    664     if regularizer is not None:
    665       # TODO(fchollet): in the future, this should be handled at the

/usr/local/lib/python3.7/dist-packages/tensorflow/python/training/tracking/base.py in _add_variable_with_custom_getter(self, name, shape, dtype, initializer, getter, overwrite, **kwargs_for_getter)
    816         dtype=dtype,
    817         initializer=initializer,
--> 818         **kwargs_for_getter)
    819 
    820     # If we set an initializer and the variable processed it, tracking will not

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer_utils.py in make_variable(name, shape, dtype, initializer, trainable, caching_device, validate_shape, constraint, use_resource, collections, synchronization, aggregation, partitioner)
    127       synchronization=synchronization,
    128       aggregation=aggregation,
--> 129       shape=variable_shape if variable_shape else None)
    130 
    131 

/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/variables.py in __call__(cls, *args, **kwargs)
    264   def __call__(cls, *args, **kwargs):
    265     if cls is VariableV1:
--> 266       return cls._variable_v1_call(*args, **kwargs)
    267     elif cls is Variable:
    268       return cls._variable_v2_call(*args, **kwargs)

/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/variables.py in _variable_v1_call(cls, initial_value, trainable, collections, validate_shape, caching_device, name, variable_def, dtype, expected_shape, import_scope, constraint, use_resource, synchronization, aggregation, shape)
    225         synchronization=synchronization,
    226         aggregation=aggregation,
--> 227         shape=shape)
    228 
    229   def _variable_v2_call(cls,

/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/variables.py in getter(**kwargs)
     65 
     66   def getter(**kwargs):
---> 67     return captured_getter(captured_previous, **kwargs)
     68 
     69   return getter

/usr/local/lib/python3.7/dist-packages/tensorflow/python/distribute/distribute_lib.py in creator_with_resource_vars(next_creator, **kwargs)
   2125         checkpoint_restore_uid = None
   2126 
-> 2127       created = self._create_variable(next_creator, **kwargs)
   2128 
   2129       if checkpoint_restore_uid is not None:

/usr/local/lib/python3.7/dist-packages/tensorflow/python/distribute/tpu_strategy.py in _create_variable(self, next_creator, **kwargs)
   1167         self._container_strategy(), _real_mirrored_creator,
   1168         distribute_utils.TPU_VARIABLE_CLASS_MAPPING,
-> 1169         distribute_utils.TPU_VARIABLE_POLICY_MAPPING, **kwargs)
   1170 
   1171   def _gather_to_implementation(self, value, destinations, axis, options):

/usr/local/lib/python3.7/dist-packages/tensorflow/python/distribute/distribute_utils.py in create_mirrored_variable(strategy, real_mirrored_creator, class_mapping, policy_mapping, **kwargs)
    306   # here.
    307   with tape.stop_recording():
--> 308     value_list = real_mirrored_creator(**kwargs)
    309     # MirroredVariable is recreated during saved_model loading, and its
    310     # component variables (value_list) will have None initializer. We

/usr/local/lib/python3.7/dist-packages/tensorflow/python/distribute/tpu_strategy.py in _real_mirrored_creator(**kwargs)
   1146             with maybe_init_scope():
   1147               initial_value = initial_value() if callable(
-> 1148                   initial_value) else initial_value
   1149 
   1150           if i > 0:

/usr/local/lib/python3.7/dist-packages/keras/initializers/initializers_v2.py in __call__(self, shape, dtype, **kwargs)
    515     else:
    516       limit = math.sqrt(3.0 * scale)
--> 517       return self._random_generator.random_uniform(shape, -limit, limit, dtype)
    518 
    519   def get_config(self):

/usr/local/lib/python3.7/dist-packages/keras/initializers/initializers_v2.py in random_uniform(self, shape, minval, maxval, dtype)
    971       op = tf.random.uniform
    972     return op(
--> 973         shape=shape, minval=minval, maxval=maxval, dtype=dtype, seed=self.seed)
    974 
    975   def truncated_normal(self, shape, mean, stddev, dtype):

/usr/local/lib/python3.7/dist-packages/tensorflow/python/util/dispatch.py in wrapper(*args, **kwargs)
    204     """Call target, and fall back on dispatchers if there is a TypeError."""
    205     try:
--> 206       return target(*args, **kwargs)
    207     except (TypeError, ValueError):
    208       # Note: convert_to_eager_tensor currently raises a ValueError, not a

/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/random_ops.py in random_uniform(shape, minval, maxval, dtype, seed, name)
    300     maxval_is_one = isinstance(maxval, int) and maxval == 1
    301     if not minval_is_zero or not maxval_is_one or dtype.is_integer:
--> 302       minval = ops.convert_to_tensor(minval, dtype=dtype, name="min")
    303       maxval = ops.convert_to_tensor(maxval, dtype=dtype, name="max")
    304     seed1, seed2 = random_seed.get_seed(seed)

/usr/local/lib/python3.7/dist-packages/tensorflow/python/profiler/trace.py in wrapped(*args, **kwargs)
    161         with Trace(trace_name, **trace_kwargs):
    162           return func(*args, **kwargs)
--> 163       return func(*args, **kwargs)
    164 
    165     return wrapped

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py in convert_to_tensor(value, dtype, name, as_ref, preferred_dtype, dtype_hint, ctx, accepted_result_types)
   1564 
   1565     if ret is None:
-> 1566       ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
   1567 
   1568     if ret is NotImplemented:

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/tensor_conversion_registry.py in _default_conversion_function(***failed resolving arguments***)
     50 def _default_conversion_function(value, dtype, name, as_ref):
     51   del as_ref  # Unused.
---> 52   return constant_op.constant(value, dtype, name=name)
     53 
     54 

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/constant_op.py in constant(value, dtype, shape, name)
    270   """
    271   return _constant_impl(value, dtype, shape, name, verify_shape=False,
--> 272                         allow_broadcast=True)
    273 
    274 

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/constant_op.py in _constant_impl(value, dtype, shape, name, verify_shape, allow_broadcast)
    281       with trace.Trace("tf.constant"):
    282         return _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
--> 283     return _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
    284 
    285   g = ops.get_default_graph()

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/constant_op.py in _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
    306 def _constant_eager_impl(ctx, value, dtype, shape, verify_shape):
    307   """Creates a constant on the current device."""
--> 308   t = convert_to_eager_tensor(value, ctx, dtype)
    309   if shape is None:
    310     return t

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/constant_op.py in convert_to_eager_tensor(value, ctx, dtype)
    104       dtype = dtypes.as_dtype(dtype).as_datatype_enum
    105   ctx.ensure_initialized()
--> 106   return ops.EagerTensor(value, ctx.device_name, dtype)
    107 
    108 

NotFoundError: '_EagerConst' is neither a type of a primitive operation nor a name of a function registered in binary running on n-c9063562-w-0. Make sure the operation or function is registered in the binary running in this process.```

**Comment by bliep** [comment]
@sushreebarsa Using the default Colab TPU environment with TF 2.5, can you reproduce the issue (60x decrease/increase in training speed) with the provided sample code?

**Comment by sushreebarsa** [comment]
@sanatmpa1  Was able to reproduce the issue (60x decrease/increase in training speed) using the default Colab TPU environment with TF 2.5 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/94852cc44f4a3dd53256f48a38178ed4/tpu-performance-issue.ipynb#scrollTo=P2EL671fL1Sg) for reference .I also tried the code on colab using latest TF v2.6.0 and got the error as mentioned in above [comment](https://github.com/tensorflow/tensorflow/issues/51485#issuecomment-898564779), please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/7415a3ea10359644eab420f41f620f3d/untitled384.ipynb) for your reference .Thank you!

**Comment by bliep** [comment]
The same issue (60x decrease/increase) of training performance also occurs on the (recently introduced) TPU-VM.
The following code is used to demonstrate the issue, uncomment the line with dummy OP to get a 60x training speed increase.

__Create a TPU-VM and connect with SSH__
```
gcloud alpha compute tpus tpu-vm create tpu_vm_test \
  --zone=europe-west4-a \
  --accelerator-type=v2-8 \
  --preemptible \
  --version=v2-alpha

gcloud alpha compute tpus tpu-vm ssh tpu_vm_test \
  --zone europe-west4-a
```

__On the TPU-VM run the following with python3:__
```
import os
import tensorflow as tf

print(f'tensorflow version {tf.__version__}')

n_batch = 1024
n_seq_len = 2048
n_bins = 64
n_nodes = 128

def create_dataset():
    ds = tf.data.Dataset.from_tensor_slices([tf.ones((n_batch, n_seq_len, n_bins))])
    def rand_batch(x):
        return tf.random.uniform((1,))*x, tf.random.uniform((1,))*x
    return ds.map(rand_batch).repeat()

def create_model():
  x = tf.keras.Input(shape=(n_seq_len, n_bins))
  g = tf.keras.layers.GRU(n_nodes, return_sequences=True)(x)
  g = tf.keras.layers.GRU(n_nodes, return_sequences=True)(g)
  g = tf.keras.layers.GRU(n_nodes, return_sequences=True)(g)
  g = tf.keras.layers.GRU(n_nodes, return_sequences=True)(g)
  # g = 1 * g # dummy OP
  g = tf.keras.layers.Dense(n_bins, activation='sigmoid')(g)
  return tf.keras.Model(inputs=[x], outputs=[g])

cluster_resolver = tf.distribute.cluster_resolver.TPUClusterResolver(tpu='local')
tf.tpu.experimental.initialize_tpu_system(cluster_resolver)
strategy = tf.distribute.TPUStrategy(cluster_resolver)
print("All devices: ", tf.config.list_logical_devices('TPU'))

with strategy.scope():

  model = create_model()
  model.summary()
  model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001), loss='mse')
  model.fit(
      create_dataset(),
      steps_per_epoch=512,
      epochs=10,
      verbose=1
  )
```

(note that the TPU-VM uses TF 2.6.0, however the code does not crash like Colab TPU 2.6.0)

**Comment by saeta** [comment]
Sorry @jvishnuvardhan , I don't work on TF anymore. Reassigning back to you to re-triage. Sorry!

**Comment by bliep** [comment]
This issue seems to be fixed in TF 2.7.0 :)
